src/service-worker/core.js

```diff
@@ -0,0 +1,306 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import '../../third_party/babel/custom-babel-helpers';
+import indexedDBP from '../../third_party/indexed-db-as-promised/index';
+
+/**
+ * The SW's current version. We conveniently use JS Date timestamps as version
+ * numbers, so we use this to determine how old a given versioned file is
+ * compared to the latest version (this version).
+ * @const
+ */
+const VERSION = '$internalRuntimeVersion$';
+
+/**
+ * Returns the version of a given versioned JS file.
+ *
+ * @param {string} url
+ * @return {string}
+ */
+function ampVersion(url) {
+  // RTVs are 2 digit prefixes followed by the current timestamp.
+  const matches = /rtv\/(\d{2}\d{13,})/.exec(url);
+  return matches ? matches[1] : VERSION;
+}
+
+/**
+ * Returns the basename (AKA the filename) of a url, used to key a url (since
+ * our JS filenames are unique).
+ *
+ * @param {string} url
+ * @return {string}
+ */
+function basename(url) {
+  return url.substr(url.lastIndexOf('/') + 1);
+}
+
+/**
+ * Returns the url with the requested version changed to `version`.
+ *
+ * @param {string} url
+ * @param {string} version
+ * @return {string}
+ */
+function versionedUrl(url, version) {
+  // Ensure we do not replace the prefix.
+  return url.replace(ampVersion(url), version);
+}
+
+/**
+ * Determines if a url is a request to a CDN JS file.
+ * We only consider "prod" (RTV 01) files.
+ * @param {string} url
+ * @return {boolean}
+ */
+function isCdnJsFile(url) {
+  return url.startsWith('https://cdn.ampproject.org/') &&
```

Comment:
I think we need to be more specific
`https://cdn.ampproject.org/rtv/` and maybe `https://cdn.ampproject.org/v0`

Otherwise you could have an HTML file cached at `https://cdn.ampproject.org/c/foo.com/doc.js`